### ISSUE TITLE
Creating Contact with non-Home Primary Address Type Results in Duplicate Address Records

### ISSUE THREAD

**Issue by jesslopez-sf** [SalesforceFoundation/NPSP#6416]
### Summary
If a customer creates a Contact and sets the Primary Address Type to something other than “Home”, NPSP will create duplicate Address records. One Address record will have an Address Type of “Home” and the other will have an Address Type of whatever was chosen on the Contact. All other information will be the same.

Likely, semi-related to Issue: #5266

### Root Cause
High-level, when a Contact is inserted, NPSP will create its Account and then construct its Address record from the Billing Address fields on the Account. There is no Primary Address Type field on the Account, so the Address_Type__c field on the Address gets set to its default picklist value, which is “Home”. Later on in the transaction, when our ADDR_Contact_TDTM Trigger Handler runs on Contact in the After Insert context, we detect a difference between the values that we would use to construct an Address from the Contact vs. what we already have in the system for the Account (the difference being the Address Type values). As a result, we insert a 2nd Address within the transaction and explicitly set its Address_Type__c to the value from the Contact (i.e. “Work”). 

The duplication does not occur if a Contact’s Primary Address Type is set to “Home” initially, since the Address created from the Account defaults to “Home”. It also does not occur if you enable the “Simple Address Change Treated as Update” option within NPSP Settings. This is because we only detect 1 change within the ADDR_Addresses_TDTM’s isAddressMatchFuzzy() method and just update the existing Address record

**More Detail:**
Within the BeforeInsert context of Contact, the  ACCT_IndividualAccounts_TDTM constructs the new Account and sets its Billing Address based on the Mailing Address values of the Contact. When we run ADDR_Account_TDTM  within the After Insert Context of the Account, we call the copyAddressStdSObjAddr() method, which takes an Address Type parameter. It’s null coming from the Account, so we set it to the default picklist value of Address_Type__c. 
```
13:30:34.0 (528842018)|VARIABLE_ASSIGNMENT|[1133]|
this.Address_Type__c|"Home"|0x46b08589
```
Then call ADDR_Addresses_TDTM.createNonDuplicateAddresses() which calls findDuplicateAddresses(). There are no duplicates, so we add that Address with an Address Type of “Home” to the dmlWrapper objectsToInsert list.
```
13:30:34.0 (673412964)|VARIABLE_ASSIGNMENT|[151]|
this.objectsToInsert|[
{"Household_Account__c":"001J000002ZzCcgIAF",
"Default_Address__c":true,
"Latest_Start_Date__c":"2021-05-13T00:00:00.000Z",
"MailingStreet__c":"123 Main Street",
"MailingCity__c":"New York","MailingPostalCode__c":"12345",
"MailingState__c":"NY","Address_Type__c":"Home"}]|0x6dcde760
```
Within the After Insert context of the Contact, we call ADDR_Contact_TDTM and start constructing an address again using copyAddressStdSObjAddress(). The Primary Address Type is not null coming from the Contact, so we set Address_Type__c to its actual value (“Work”). 
```
13:30:41.764 (8157961148)|VARIABLE_ASSIGNMENT|[1128]|this.Address_Type__c|"Work"|0x6eab3f88
```
Next we call findDuplicateAddresses() and query for existing Addresses, which returns the committed Address where Address_Type__c = home. We call isAddressChanged() and don’t find an exact match due to mismatch of address type. No fuzzy match found since Simple Address option is not enabled. We add the newly constructed Address to mapAddrDupAddr and return. We then hit lines 468-472 in createAddrFromContact() method:
https://github.com/SalesforceFoundation/NPSP/blob/71e68e262b2a29d598393aa6e26be0ac76bae02f/src/classes/ADDR_Contact_TDTM.cls#L468-L472
So we set the new constructed address to the Default_Address__c and add it to listAddrInsertDefault. We then add that address to the dmlWrapper’s objectsToInsert list and the duplicate is created.
```
13:30:41.764 (8175449768)|VARIABLE_ASSIGNMENT|[151]|this.objectsToInsert|[{"Household_Account__c":"001J000002ZzCcgIAF","Default_Address__c":true,"Latest_Start_Date__c":"2021-05-13T00:00:00.000Z","MailingStreet__c":"123 Main Street","MailingCity__c":"New York","MailingPostalCode__c":"12345","MailingState__c":"NY","Address_Type__c":"Work"}]|0x3440b60f
```

### Steps to Repeat
1. [Bug]: Create a Contact and enter First/Last Names, Primary Address Type = “Work”, Mailing Street = “123 1st Avenue”, Mailing City = “New York”, Mailing State = “NY”, Mailing Postal Code = “12345”. Save. Confirm a Contact and Account are created, along with 2 Address records.
2. [Works as expected]: Create a Contact and enter First/Last Names, Primary Address Type = “Home”, Mailing Street = “123 1st Avenue”, Mailing City = “New York”, Mailing State = “NY”, Mailing Postal Code = “12345”. Save. Confirm a Contact and Account are created, along with just 1 Address record.
3. [Works as expected]: Go to NPSP Settings → People → Addresses and enable the “Simple Address Change Treated as Update” option. Create a Contact and enter First/Last Names, Primary Address Type = “Work”, Mailing Street = “123 1st Avenue”, Mailing City = “New York”, Mailing State = “NY”, Mailing Postal Code = “12345”. Save. Confirm a Contact and Account are created, along with just 1 Address record.

### Workaround
Go to NPSP Settings → People → Addresses and enable the “Simple Address Change Treated as Update” option.

**Comment by jesslopez-sf** [comment]
W-9275840

**Comment by jesslopez-sf** [comment]
This is resolved in NPSP 3.201
